### DISCUSSION TITLE
Can't get the message from the LSB Image in PDF

### DISCUSSION THREAD

**Discussion by roylo97** [pymupdf/PyMuPDF#2729] [Looking for help]
I would like to add message in image of PDF. Such as LSB.

Most of time, it can successful to update the image in pdf with LSB.

but then the image with LSB extracted from the PDF, the image seems like changed (compare with their raw data) and I can't read the LSB message.

Am I misunderstood that inputed image == exported image?

**Comment by JorjMcKie** [comment]
I don't understand: what is LSB?
How do you add "message in image"?

**Reply by roylo97** [reply]
Sorry for the late reply.

LSB is steganography which hides messages inside an image by replacing the Least significant bit of the image with the bits of a message to be hidden.

I would like to put some copyright message in the image instead of a watermark

When I tried to extract the message, there are failed.

So, the question is
is it some conversion when there is inserting/replacing an image, so that means it will lose some data

Here are my testing case and my simple code
![conceptual](https://github.com/pymupdf/PyMuPDF/assets/71375868/e0660940-9a8e-4698-ab63-5e406998e4a3)
[test-pymupdf.zip](https://github.com/pymupdf/PyMuPDF/files/12914456/test-pymupdf.zip)

**Comment by JorjMcKie** [comment]
The image extracted via `doc.extract_image(xref)` should deliver the original image - more or less.
Deviations always occur if the image in the PDF is transparent. Then a separate mask image exists.
In that case, the original image can only be recovered via overlaying the two pixmaps that correspond to the based image, respectively the mask image.
Deviations also **_always_** occur if you build the image via `pix = fitz.Pixmap(doc, xref)`. This approach ignores the original image format, and only those image outputs as supported by `Pixmap` are available.

Binary image equality can only be hoped for when looking at `img["image"]` when `img = doc.extract_image(xref)`.
